### PR TITLE
Include Static Spec Test Vectors for SSZ

### LINKED ISSUE #22: BeaconBlockBody Marshal Fails
This is a tracking issue for a particular problem encountered when marshaling a beacon block body type as generated by yaml go.

## Background On How SSZ Works

SSZ is a straightforward approach towards taking an arbitrary input and producing a slice of bytes from it. Given input can take the form of structs and lists which may have elements of variable size, **and the output is a flattened slice of bytes**, we need some way to determine item delimiters or indices at which variable size elements/fixed size elements occur. Let's define some terms first:

A type is said to be **fixed size** if it:
- Is a basic type, such as bool, uint8 - uint64
- Is an array of fixed sized elements
- Is a struct where, recursively, every field is fixed size

A type is said to be **variable size** if it:
- Is a slice
- Is an array where at least one inner element is variable size
- Is a struct where, recursively, at least one element is variable size

Let's see some examples:

- `[32][]byte{}` is considered variable size
- `[4]byte{
  SomeItem struct {
    Slot uint64
    Root [32]byte
  }
}` is considered fixed size
- `AnotherItem struct { CustodyBitfield []byte Slot uint32 }` is considered variable size

and finally
```go
type Item struct {
  Root [32]byte
  Slot uint64
  NestedItem struct {
    Slot uint64
  }
}
```
is considered fixed size and
```go
type Item struct {
  Root [32]byte
  Slot uint64
  NestedItem struct {
    Slot uint64
    SomeField [32][]uint64
  }
}
```
is considered variable size.

### How Offsets Work

Consider the type `[][]uint16{{1, 2}, {3}}`, it will marshal into
```
[8 0 0 0 12 0 0 0 1 0 2 0 3 0]
```
This basically says that the first element begins at index `8`, and the next element begins at index `12`, given we need a way to determine when the variable size parts of marshaled bytes begin.

Offsets **do not** always happen at the beginning of the marshaled result, as they can be interwoven with fixed size elements. Consider the struct:
```go
type Item struct {
  ValidatorIndex uint64
  NestedItem struct {
    SomeField []byte
    Slot uint64
  }
}
```
The marshaled results will look something along the lines of

```
[ValidatorIndex, offset where SomeField begins, SomeField, slot]
```

### How We Determine Offsets

All of our marshal functions return the size of the element we are marshaling. This will help us track the next offset. For example, say we are serializing:
```go
type Item struct {
  ValidatorIndex uint64
  NestedItem struct {
    SomeField []byte
    Slot uint64
  }
}
```
We first compute the total size of the item. Let's assume that SomeField has 3 elements in it. Then, the total size of the struct is 8 (for uint64) + (3 + offset) + 8. We use this to preallocate a buffer we will write into during the marshaling process. 

We set a `current writing index = 0`, and then marshal validator index, which returns a size of `8`, we update the `current writing index  += 8`. We then determine that SomeField is variable size, so we will write an offset for it starting at index 8 (we will write 4 bytes as offsets are uint32). Then, we marshal the byte slice which returns length 3, so we update `current index += 4 + 3`. Then we serialize the Slot field, which has size 8 so we update `current index += 8` and we are done. The same procedure applies for all sorts of complex types.

## The Problem

```go
BeaconBlockBody struct {
			Value struct {
				RandaoReveal []byte `json:"randao_reveal" ssz:"size=96"`
				Eth1Data     struct {
					DepositRoot  []byte `json:"deposit_root" ssz:"size=32"`
					DepositCount uint64 `json:"deposit_count"`
					BlockHash    []byte `json:"block_hash" ssz:"size=32"`
				} `json:"eth1_data"`
				Graffiti          []byte `json:"graffiti" ssz:"size=32"`
				ProposerSlashings []struct {
					ProposerIndex uint64 `json:"proposer_index"`
					Header1       struct {
						Slot       uint64 `json:"slot"`
						ParentRoot []byte `json:"parent_root" ssz:"size=32"`
						StateRoot  []byte `json:"state_root" ssz:"size=32"`
						BodyRoot   []byte `json:"body_root" ssz:"size=32"`
						Signature  []byte `json:"signature" ssz:"size=96"`
					} `json:"header_1"`
					Header2 struct {
						Slot       uint64 `json:"slot"`
						ParentRoot []byte `json:"parent_root" ssz:"size=32"`
						StateRoot  []byte `json:"state_root" ssz:"size=32"`
						BodyRoot   []byte `json:"body_root" ssz:"size=32"`
						Signature  []byte `json:"signature" ssz:"size=96"`
					} `json:"header_2"`
				} `json:"proposer_slashings"`
				AttesterSlashings []struct {
					Attestation1 struct {
						CustodyBit0Indices []uint64 `json:"custody_bit_0_indices"`
						CustodyBit1Indices []uint64 `json:"custody_bit_1_indices"`
						Data               struct {
							BeaconBlockRoot []byte `json:"beacon_block_root" ssz:"size=32"`
							SourceEpoch     uint64 `json:"source_epoch"`
							SourceRoot      []byte `json:"source_root" ssz:"size=32"`
							TargetEpoch     uint64 `json:"target_epoch"`
							TargetRoot      []byte `json:"target_root" ssz:"size=32"`
							Crosslink       struct {
								Shard      uint64 `json:"shard"`
								StartEpoch uint64 `json:"start_epoch"`
								EndEpoch   uint64 `json:"end_epoch"`
								ParentRoot []byte `json:"parent_root" ssz:"size=32"`
								DataRoot   []byte `json:"data_root" ssz:"size=32"`
							} `json:"crosslink"`
						} `json:"data"`
						Signature []byte `json:"signature" ssz:"size=96"`
					} `json:"attestation_1"`
					Attestation2 struct {
						CustodyBit0Indices []uint64 `json:"custody_bit_0_indices"`
						CustodyBit1Indices []uint64 `json:"custody_bit_1_indices"`
						Data               struct {
							BeaconBlockRoot []byte `json:"beacon_block_root" ssz:"size=32"`
							SourceEpoch     uint64 `json:"source_epoch"`
							SourceRoot      []byte `json:"source_root" ssz:"size=32"`
							TargetEpoch     uint64 `json:"target_epoch"`
							TargetRoot      []byte `json:"target_root" ssz:"size=32"`
							Crosslink       struct {
								Shard      uint64 `json:"shard"`
								StartEpoch uint64 `json:"start_epoch"`
								EndEpoch   uint64 `json:"end_epoch"`
								ParentRoot []byte `json:"parent_root" ssz:"size=32"`
								DataRoot   []byte `json:"data_root" ssz:"size=32"`
							} `json:"crosslink"`
						} `json:"data"`
						Signature []byte `json:"signature" ssz:"size=96"`
					} `json:"attestation_2"`
				} `json:"attester_slashings"`
				Attestations []struct {
					AggregationBitfield []byte `json:"aggregation_bitfield"`
					Data                struct {
						BeaconBlockRoot []byte `json:"beacon_block_root" ssz:"size=32"`
						SourceEpoch     uint64 `json:"source_epoch"`
						SourceRoot      []byte `json:"source_root" ssz:"size=32"`
						TargetEpoch     uint64 `json:"target_epoch"`
						TargetRoot      []byte `json:"target_root" ssz:"size=32"`
						Crosslink       struct {
							Shard      uint64 `json:"shard"`
							StartEpoch uint64 `json:"start_epoch"`
							EndEpoch   uint64 `json:"end_epoch"`
							ParentRoot []byte `json:"parent_root" ssz:"size=32"`
							DataRoot   []byte `json:"data_root" ssz:"size=32"`
						} `json:"crosslink"`
					} `json:"data"`
					CustodyBitfield []byte `json:"custody_bitfield"`
					Signature       []byte `json:"signature" ssz:"size=96"`
				} `json:"attestations"`
				Deposits []struct {
					Proof [][]byte `json:"proof" ssz:"size=32,32"`
					Data  struct {
						Pubkey                []byte `json:"pubkey" ssz:"size=48"`
						WithdrawalCredentials []byte `json:"withdrawal_credentials" ssz:"size=32"`
						Amount                uint64 `json:"amount"`
						Signature             []byte `json:"signature" ssz:"size=96"`
					} `json:"data"`
				} `json:"deposits"`
				VoluntaryExits []struct {
					Epoch          uint64 `json:"epoch"`
					ValidatorIndex uint64 `json:"validator_index"`
					Signature      []byte `json:"signature" ssz:"size=96"`
				} `json:"voluntary_exits"`
				Transfers []struct {
					Sender    uint64 `json:"sender"`
					Recipient uint64 `json:"recipient"`
					Amount    uint64 `json:"amount"`
					Fee       uint64 `json:"fee"`
					Slot      uint64 `json:"slot"`
					Pubkey    []byte `json:"pubkey" ssz:"size=48"`
					Signature []byte `json:"signature" ssz:"size=96"`
				} `json:"transfers"`
			} `json:"value"`
			Serialized []byte `json:"serialized"`
			Root       []byte `json:"root" ssz:"size=32"`
		} `json:"BeaconBlockBody,omitempty"`
```

In order to reproduce, checkout the `core-spec-test` locally, navigate to the `spectests` directory, and run `go test .`. The error message is:

```
--- FAIL: TestYaml (0.20s)
panic: runtime error: slice bounds out of range [recovered]
	panic: runtime error: slice bounds out of range

goroutine 36 [running]:
testing.tRunner.func1(0xc000144100)
	/usr/local/go/src/testing/testing.go:792 +0x387
panic(0x13e7c80, 0x177ea80)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/prysmaticlabs/go-ssz.marshalByteArray(0x13bd4a0, 0xc00126ae00, 0x197, 0xc0007f0000, 0xc1a, 0xc1a, 0xbca, 0xbca, 0x0, 0x0)
```

This occur when marshaling the very last item of the beacon block body, which is the `Signature` of the `Transfers` slice. When marshaling SSZ, we preallocate a bytes buffer of a certain size that we will write into for each field. We have a `size.go` file that recursively determines how much we should allocate based on the size of every value within the input. **This is confirmed to be correct and match the length of the serialized output (3098 == 3098)**.

The last bytes of the _expected_ output are:
```
... 61 113 55 224 117 108 211 134 98 146 124 185 89 161 212 98 220 247 168 37 111 33 206 114 171 141 150 253 180 155 190 162 183 0 103 95 235 228 35 224 131 224 254 140 154 224 78 33 193 85 115 128 36 125 51 78 20 88 183 98 135 108 228 62 33 124 115 126 62 84 117 250 32 79 42 89 167 52 154 236 133 181 236 119 190 153 252 151 113 199 245 221 69 22 191 78 77 13 151 3 117 255 20 247 24 42 2]
```
But our buffer right before the panic looks like:
```
... 61 113 55 224 117 108 211 134 98 146 124 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
```

Which looks like it is missing 16 extra bytes to fit in the signature ([96]byte) as there are only 80 zeros.

Putting the text through [diffchecker.com](https://www.diffchecker.com/) shows us that we add several extra offsets (4 to be exact), right before the `[]struct` fields such as `[]ProposerSlashing`, `[]Transfers`, etc.

**Left: Expected, Right: Our Output**
Screen Shot 2019-06-19 at 4 05 57 PM

Screen Shot 2019-06-19 at 4 06 05 PM

The primary suspicion is something is being added when are determining if a struct field is fixed or variable size in the struct marshaler found in `marshal.go`. There is a condition in that function that goes as follows:

```go
	for i, f := range fields {
			if !isVariableSizeType(val.Field(i), f.typ) {
				fixedIndex, err = f.sszUtils.marshaler(val.Field(i), buf, fixedIndex)
				if err != nil {
					return 0, err
				}
			} else {
				fmt.Printf("Writing %v\n", f.name)
				nextOffsetIndex, err = f.sszUtils.marshaler(val.Field(f.index), buf, currentOffsetIndex)
				if err != nil {
					return 0, err
				}
				// Write the offset.
				offsetBuf := make([]byte, BytesPerLengthOffset)
				binary.LittleEndian.PutUint32(offsetBuf, uint32(currentOffsetIndex-startOffset))
				copy(buf[fixedIndex:fixedIndex+uint64(BytesPerLengthOffset)], offsetBuf)

				// We increase the offset indices accordingly.
				currentOffsetIndex = nextOffsetIndex
				fixedIndex += uint64(BytesPerLengthOffset)
			}
```
We keep track of the last index written for fixed size elements and for variable size. elements
it basically says, if a field is not variable size type, we marshal it normally and update the `fixedIndex` variable to be the last written index of that marshaling event. If it is of variable size, we write an offset accordingly.

My suspicion is that block operations such as `[]Transfer`, etc. should not be treated as variable size since they have a `MAX_TRANSFERS` limiting value, which might be causing some of these problems. **However**, initial conversations with Cayman from Chainsafe have led me to believe that they should indeed be treated as variable elements.

The goal of this tracking issue is to determine other ways of debugging these problems and getting to the bottom of this small diff that prevents our tests from passing.

### References

- [Simple Serialize Spec](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/simple-serialize.md)
- [Chainsafe's Marshal Functions](https://github.com/ChainSafe/ssz-js/blob/28282b64c68f19a250c920e66ae6d61fbef72219/src/serialize.ts#L90)

### PR DESCRIPTION
**Pull request by rauljordan**
## Background

This resolves #22. SSZ Generic tests will be included in a separate PR.

This PR includes all the necessary tests as of the v0.7.1 release of the official Ethereum 2.0 specification for simple serialize. Additionally, this PR supports marshaling and umarshaling of protobuf structs and all other structs which do not have the ability to specify array type (fixed-size) fields in their definitions. That is, one can allow a slice type field to be treated as fixed size by adding a \`ssz:"size=32"\` struct tag. For example:

```go
type SomeItem struct {
  BlockRoot []byte `ssz:"size=32"`
}
```
This will allow the marshaler to treat the `BlockRoot` field of the struct as a fixed-size, `[32]byte` array. For multidimensional arrays, sizes can be specified by comma-separating the size values:

```go
type SomeItem struct {
  MerkleProof [][]byte `ssz:"size=32,32"`
}
```

For items in which one dimension is unbounded, the unbounded dimension can be specified with a question mark `?`:

```go
type SomeItem struct {
  HistoricalBlockRoots [][]byte `ssz:"size=?,32"`
}
```

This will treat the field as `[][32]byte` when marshaling.

## Spec Tests

All tests pass for both mainnet and minimal configuration options of data structures in Eth2:

```
$ go test . -v                                                                                                                                                                  2.6.0
=== RUN   TestYamlSpecTests
--- PASS: TestYamlSpecTests (5.03s)
    ssz_spec_test.go:27: Running spec tests for: ssz_bench_block.yaml
    ssz_spec_test.go:27: Running spec tests for: ssz_bench_state.yaml
    ssz_spec_test.go:27: Running spec tests for: ssz_mainnet_random_formatted.yaml
    ssz_spec_test.go:27: Running spec tests for: ssz_minimal_lengthy_formatted.yaml
    ssz_spec_test.go:27: Running spec tests for: ssz_minimal_max_formatted.yaml
    ssz_spec_test.go:27: Running spec tests for: ssz_minimal_nil_formatted.yaml
    ssz_spec_test.go:27: Running spec tests for: ssz_minimal_one_formatted.yaml
    ssz_spec_test.go:27: Running spec tests for: ssz_minimal_random_chaos_formatted.yaml
    ssz_spec_test.go:27: Running spec tests for: ssz_minimal_random_formatted.yaml
    ssz_spec_test.go:27: Running spec tests for: ssz_minimal_zero_formatted.yaml
PASS
ok      github.com/prysmaticlabs/go-ssz/spectests       5.080s
```

## Benchmarks
Initial benchmarks for raw, unoptimized, but **correct** SSZ are as follows

```
$ go test -bench=. -benchmem                                                                                                                                                    2.6.0
goos: darwin
goarch: amd64
pkg: github.com/prysmaticlabs/go-ssz/spectests
BenchmarkMainnetBeaconBlock_Marshal-4               2000            577005 ns/op           60344 B/op       3312 allocs/op
BenchmarkMainnetBeaconBlock_HashTreeRoot-4          2000            919728 ns/op          352672 B/op       6586 allocs/op
BenchmarkMainnetBeaconBlock_SigningRoot-4           2000            935323 ns/op          350704 B/op       6557 allocs/op
BenchmarkMainnetBeaconState_Marshal-4                100          21182819 ns/op         2859073 B/op      75143 allocs/op
BenchmarkMainnetBeaconState_HashTreeRoot-4            20          62647973 ns/op        25509234 B/op     382567 allocs/op
```

This is using the mainnet, full beacon state and blocks from a yaml file. There is a lot of room for optimization and improvement of the codebase but that will be out of scope of this PR.

### CODE REVIEW

File: spectests/ssz_spec_test.go [hunk 1]
```diff
@@ -0,0 +1,811 @@
...
+	"strings"
+	"testing"
+
+	"github.com/bazelbuild/rules_go/go/tools/bazel"
+	"github.com/ghodss/yaml"
+	"github.com/prysmaticlabs/go-ssz"
+)
+
+func TestYamlSpecTests(t *testing.T) {
+	topPath := "/eth2_spec_tests/tests/ssz_static/core/"
```
Comment (prestonvanloon) [inline spectests/ssz_spec_test.go]: This is only the core files? What about ssz_generic/uint? Maybe for another PR, but we are missing some spec tests.

File: spectests/BUILD.bazel [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    testonly = True,
+    srcs = [
+        "ssz_mainnet.yaml.go",
+        "ssz_minimal.yaml.go",
+    ],
+    importpath = "github.com/prysmaticlabs/go-ssz/spectests",
+    visibility = ["//visibility:public"],
+)
+
+go_test(
```
Comment (terencechain) [inline spectests/BUILD.bazel]: don't forget to add the test size and tag spec tests with `manual` and `spectest`

Comment (prestonvanloon) [reply spectests/BUILD.bazel]: These aren't as expensive as I expected, we can omit `manual` but definitely add `spectest`

File: WORKSPACE [hunk 3]
```diff
@@ -28,6 +28,21 @@ load("@com_github_prysmaticlabs_go_ssz//:deps.bzl", "go_ssz_dependencies")
...
+    build_file_content = """
+filegroup(
+    name = "test_data",
+    srcs = glob([
+        "**/*.yaml",
+    ]),
+    visibility = ["//visibility:public"],
+)
+    """,
+    sha256 = "d297d5df03d1a635450cdb6d746e585485529ce13f121fffc375ad12470ae32f",
```
Comment (prestonvanloon) [inline WORKSPACE]:
```suggestion
    sha256 = "56847989737e816ab7d23f3bb2422347dfa81271bae81a94de512c01461fab25",
```

File: struct_utils.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+)
+
+// We use a question mark in ssz struct tags to specify a ssz field should have
+// unbounded size, which is useful when describing slices of arrays such as [][32]byte.
+// The ssz struct tag for such field type would be `ssz:"size=?,32"`
+var unboundedSSZFieldSizeMarker = "?"
```
Comment (prestonvanloon) [inline struct_utils.go]: Should this be public for godocs?

**Pull request merged**